pkg/api/v1/state.go

```diff
@@ -0,0 +1,232 @@
+/*
+Copyright 2021 The TestGrid Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package v1 (api/v1) is the first versioned implementation of the API
+package v1
+
+import (
+	"compress/zlib"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math"
+	"net/http"
+	"net/url"
+	"path"
+	"time"
+
+	"github.com/GoogleCloudPlatform/testgrid/config"
+	apipb "github.com/GoogleCloudPlatform/testgrid/pb/api/v1"
+	configpb "github.com/GoogleCloudPlatform/testgrid/pb/config"
+	statepb "github.com/GoogleCloudPlatform/testgrid/pb/state"
+	"github.com/GoogleCloudPlatform/testgrid/util/gcs"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/gorilla/mux"
+)
+
+// gridReader returns the grid content and metadata (last updated time, generation id)
+type gridReader func(ctx context.Context) (io.ReadCloser, time.Time, int64, error)
+
+// groupFinder returns the named group as well as reader for the grid state
+type groupFinder func(string) (*configpb.TestGroup, gridReader, error)
+
+// readGrid downloads and deserializes the current test group state.
+func readGrid(ctx context.Context, reader gridReader) (*statepb.Grid, time.Time, int64, error) {
+	var t time.Time
+	r, mod, gen, err := reader(ctx)
+	if err != nil {
+		return nil, t, 0, fmt.Errorf("open: %w", err)
+	}
+	defer r.Close()
+	zlibReader, err := zlib.NewReader(r)
+	if err != nil {
+		return nil, t, 0, fmt.Errorf("decompress: %v", err)
+	}
+	buf, err := ioutil.ReadAll(zlibReader)
+	if err != nil {
+		return nil, t, 0, fmt.Errorf("read: %v", err)
+	}
+	var g statepb.Grid
+	if err = proto.Unmarshal(buf, &g); err != nil {
+		return nil, t, 0, fmt.Errorf("parse: %v", err)
+	}
+	return &g, mod, gen, nil
+}
+
+// pathReader returns a reader for the specified path and last modified, generation metadata.
+func pathReader(ctx context.Context, client gcs.Client, path gcs.Path) (io.ReadCloser, time.Time, int64, error) {
+	r, attrs, err := client.Open(ctx, path)
+	if err != nil {
+		return nil, time.Time{}, 0, fmt.Errorf("client.Open(): %w", err)
+	}
+	if attrs == nil {
+		return r, time.Time{}, 0, nil
+	}
+	return r, attrs.LastModified, attrs.Generation, nil
+}
+
+// getDashboardTabInternal locate dashboard tab in config
```

Comment:
nit: Don't think 'internal' is needed. 'findDashboardTab' maybe?

also nit: 'locates'?